feat: Enable HBase to Cloud Bigtable replication

vermas2012 opened the pull request:
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/java-bigtable-hbase/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> ☕️

igorbernstein2 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/CloudBigtableReplicator.java:
```diff
@@ -0,0 +1,305 @@
+   *
+   * @param walEntriesByTable Map of WALEntries keyed by table name.
+   */
+  public boolean replicate(Map<String, List<BigtableWALEntry>> walEntriesByTable) {
```
can you extract the per table code into a function, it would make it a lot easier to read:

for(Entry e : walEntriesByTable.entrySet()) {
   futures.add(replicateTable(e.getKey(), e.getValue()))
}
for(Future f : futures) {
  succeeded = f.get() && succeeded;
  ..
}

igorbernstein2 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/ApproximatingIncompatibleMutationAdapter.java:
```diff
@@ -0,0 +1,102 @@
+ * Practically, converts DeleteFamilyBeforeTimestamp to DeleteFamily when delete is requested before
+ * "now".
+ */
+public class ApproximatingIncompatibleMutationAdapter extends IncompatibleMutationAdapter {
```
Internal?

igorbernstein2 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/ApproximatingIncompatibleMutationAdapter.java:
```diff
@@ -0,0 +1,102 @@
+   * DeleteFamilyBeforeTimestamp where (timestamp < walkey.writeTime()) as DeleteFamily.
+   */
+  public static final String DELETE_FAMILY_WRITE_THRESHOLD_KEY =
+      "google.bigtable.deletefamily.threshold";
```
Please centralize all of the config keys

igorbernstein2 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/BigtableWALEntry.java:
```diff
@@ -0,0 +1,48 @@
+ * BigtableWALEntry abstracts minimal functionality from WAL.Entry required for this replication
+ * library.
+ */
+public class BigtableWALEntry {
```
Internal?

igorbernstein2 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/BigtableWALEntry.java:
```diff
@@ -0,0 +1,48 @@
+ * library.
+ */
+public class BigtableWALEntry {
+  private long walWriteTimestamp;
+  private ArrayList<Cell> cells;
+  private String tableName;
+
```
final?

igorbernstein2 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapter.java:
```diff
@@ -0,0 +1,148 @@
+  private final Configuration conf;
+  private final MetricsExporter metricsExporter;
+
+  public static final String INCOMPATIBLE_MUTATION_METRIC_KEY = "bigtableIncompatibleMutations";
```
Consider moving all of the metrics behind a a class so that they are centralized and typed

igorbernstein2 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapter.java:
```diff
@@ -0,0 +1,148 @@
+        } catch (UnsupportedOperationException use) {
+          // Drop the mutation, not dropping it will lead to stalling of replication.
+          incrementDroppedIncompatibleMutations();
+          LOG.error("DROPPING INCOMPATIBLE MUTATION: " + cell);
```
The error severity should be enough, no need to yell using caps

igorbernstein2 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapterFactory.java:
```diff
@@ -0,0 +1,43 @@
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.client.Connection;
+
+public class IncompatibleMutationAdapterFactory {
```
Internal?

igorbernstein2 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/metrics/MetricsExporter.java:
```diff
@@ -0,0 +1,25 @@
+ * MetricsExporter is exposed as an interface to remove dependency on MetricsSource which is in
+ * hbase-server.
+ */
+public interface MetricsExporter {
```
Internal?

vermas2012 replied:
see my comment on the 1.x package.

igorbernstein2 commented on hbase-migration-tools/bigtable-hbase-replication/pom.xml:
```diff
@@ -0,0 +1,57 @@
+    <module>bigtable-hbase-2.x-replication</module>
+  </modules>
+
+  <build>
+    <pluginManagement>
+      <plugins>
+        <plugin>
+          <groupId>org.apache.maven.plugins</groupId>
+          <artifactId>maven-javadoc-plugin</artifactId>
+          <version>3.3.1</version>
+          <executions>
+            <execution>
+              <id>default-cli</id>
+            </execution>
+          </executions>
+        </plugin>
+      </plugins>
+    </pluginManagement>
+  </build>
```
why is this necessary?

vermas2012 replied:
I don't think it is. I copied POMs from other modules. I missed it.

igorbernstein2 commented on hbase-migration-tools/pom.xml:
```diff
@@ -0,0 +1,55 @@
+    <module>bigtable-hbase-replication</module>
+  </modules>
+
+  <build>
+    <pluginManagement>
+      <plugins>
+        <plugin>
+          <groupId>org.apache.maven.plugins</groupId>
+          <artifactId>maven-javadoc-plugin</artifactId>
+          <version>3.3.1</version>
+          <executions>
+            <execution>
+              <id>default-cli</id>
+            </execution>
+          </executions>
+        </plugin>
+      </plugins>
+    </pluginManagement>
+  </build>
```
why is this necessary?

igorbernstein2 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/test/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapterTest.java:
```diff
@@ -0,0 +1,238 @@
+
+  @Mock MetricsExporter metricsExporter;
+
+  @Mock BigtableWALEntry mockWalEntry;
```
Why mock a value class? just instantiate it

vermas2012 replied:
yes, we were using HBase WAL before and there was no good way to instantiate it. 

Done.

igorbernstein2 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/test/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapterTest.java:
```diff
@@ -0,0 +1,238 @@
+
+  @Rule public final MockitoRule mockitoRule = MockitoJUnit.rule();
+
+  Configuration conf;
```
private?

igorbernstein2 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/test/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapterTest.java:
```diff
@@ -0,0 +1,238 @@
+@RunWith(JUnit4.class)
+public class IncompatibleMutationAdapterTest {
+
+  class TestIncompatibleMutationAdapter extends IncompatibleMutationAdapter {
```
static?

igorbernstein2 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/test/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapterTest.java:
```diff
@@ -0,0 +1,238 @@
+
+import com.google.cloud.bigtable.hbase.replication.metrics.MetricsExporter;
+import java.nio.charset.StandardCharsets;
+import java.util.*;
```
no wildcards please

igorbernstein2 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/test/java/com/google/cloud/bigtable/hbase/replication/adapters/ApproximatingIncompatibleMutationAdapterTest.java:
```diff
@@ -0,0 +1,139 @@
+  public void setUp() throws Exception {
+    conf = HBaseConfiguration.create();
+    conf.setInt(DELETE_FAMILY_WRITE_THRESHOLD_KEY, 10);
+    // when(conf.getInt(anyString(), anyInt())).thenReturn(10);
```
here and everywhere else please remove stale code

igorbernstein2 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/test/java/com/google/cloud/bigtable/hbase/replication/adapters/ApproximatingIncompatibleMutationAdapterTest.java:
```diff
@@ -0,0 +1,139 @@
+
+  @Before
+  public void setUp() throws Exception {
+    conf = HBaseConfiguration.create();
```
Does this need to be a n HBaseConfiguration and do you want it to search for hbase-site.xml? if not then this should be `new Configuration(false)`

vermas2012 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-1.x-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/metrics/HBaseMetricsExporter.java:
```diff
@@ -0,0 +1,34 @@
+import org.apache.hadoop.hbase.replication.regionserver.MetricsSource;
+
+/** HBaseMetricsExporter implements MetricExporter which bridges with MetricsSource. */
+public class HBaseMetricsExporter implements MetricsExporter {
```
I would like to leave it as an external API. If customers want to plug-in CBT metrics into some other system they can pass their own implementation here? Maybe?

vermas2012 replied:
actually never mind. we can make it internal

vermas2012 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-1.x-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpoint.java:
```diff
@@ -0,0 +1,88 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.hbase.replication;
```
If we have the same classpath for all endpoints, it makes the scripts and commands easier to show. Otherwise, we need to share different classpaths and different commands for enabling replication on different version. 

Not terrible, but confusing for the customers. Can you give some more details about these future modules?

vermas2012 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-1.x-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpoint.java:
```diff
@@ -0,0 +1,88 @@
+  @Override
+  protected synchronized void doStart() {
+    LOG.error(
+        "Starting replication to CBT. ", new RuntimeException("Dummy exception for stacktrace."));
```
yes. removed it.

vermas2012 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-1.x-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpoint.java:
```diff
@@ -0,0 +1,88 @@
+
+  @Override
+  public boolean replicate(ReplicateContext replicateContext) {
+    Map<String, List<BigtableWALEntry>> walEntriesByTable = new HashMap<>();
```
yeah we had it like that before. In my previous teams, we didn't prefer streams and lambda on performance critical code paths.  I have not followed the latest java release. When moving to loops, we did create a bug (b/218514410) to  do some micro benchmarks and use streams if they are performant.

vermas2012 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-1.x-replication/src/test/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpointTest.java:
```diff
@@ -0,0 +1,351 @@
+
+  @Before
+  public void setUp() throws Exception {
+    // TODO: Disable all logging for HDFS and HBase micro cluser
```
It was a reminder for me, but I did remove all the noisy logs. Removing it.

vermas2012 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-1.x-replication/src/test/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpointTest.java:
```diff
@@ -0,0 +1,351 @@
+    // TODO: Disable all logging for HDFS and HBase micro cluser
+    // Prepare HBase mini cluster configuration
+    Configuration conf = hbaseTestingUtil.getConfiguration();
+    conf.setFloat("hbase.regionserver.logroll.multiplier", 0.0003f);
```
This was supposed to be cleaned up. Sorry. I copied some configs from HBase tests, I don't think we need most of it.

vermas2012 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-1.x-replication/src/test/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpointTest.java:
```diff
@@ -0,0 +1,351 @@
+  @After
+  public void tearDown() throws Exception {
+    replicationAdmin.close();
+    hbaseTestingUtil.shutdownMiniCluster();
```
good point. closed.

vermas2012 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-1.x-replication/src/test/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpointTest.java:
```diff
@@ -0,0 +1,351 @@
+
+    // Wait for replication to catch up
+    // TODO Find a better alternative than sleeping? Maybe disable replication or turnoff mini
+    // cluster
```
I will move to utility and leave a TODO for now. We didn't have the metricsExporter when I wrote this code, now maybe we can inject a mock there and wait for all the mutations to come with a timeout. I will unblock the customer first, then address this before release.

vermas2012 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-2.x-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpoint.java:
```diff
@@ -0,0 +1,87 @@
+package com.google.cloud.bigtable.hbase.replication;
+
+import com.google.cloud.bigtable.hbase.replication.adapters.BigtableWALEntry;
+import java.util.*;
```
Really sorry about this, new dev does not have formatter set. Will get that fixed.

vermas2012 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-2.x-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpoint.java:
```diff
@@ -0,0 +1,87 @@
+
+  @Override
+  public boolean replicate(ReplicateContext replicateContext) {
+    Map<String, List<BigtableWALEntry>> walEntriesByTable = new HashMap<>();
+    for (WAL.Entry wal : replicateContext.getEntries()) {
+      String tableName = wal.getKey().getTableName().getNameAsString();
+      BigtableWALEntry bigtableWALEntry =
+          new BigtableWALEntry(wal.getKey().getWriteTime(), wal.getEdit().getCells(), tableName);
+      if (!walEntriesByTable.containsKey(tableName)) {
+        walEntriesByTable.put(tableName, new ArrayList<>());
+      }
+      walEntriesByTable.get(tableName).add(bigtableWALEntry);
+    }
+    return cloudBigtableReplicator.replicate(walEntriesByTable);
```
Lets keep the conversation in 1.x. Whatever we decide we will apply on both packages.

vermas2012 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-2.x-replication/src/test/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpointTest.java:
```diff
@@ -0,0 +1,348 @@
+// import org.slf4j.Logger;
+// import org.slf4j.LoggerFactory;
+//
+//// TODO(add serial replication tests)
```
Sushan will uncomment this, the testing util is deprecated in 2.0 without an alternative. he is looking for alternatives. Meanwhile, I prefer to have these tests run as is on deprecated utils. I have uncommented these.

vermas2012 replied:
@sushanb

vermas2012 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-2.x-replication/pom.xml:
```diff
@@ -0,0 +1,224 @@
+
+            <plugin>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <version>2.5.1</version>
```
ack.

vermas2012 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/CloudBigtableReplicationTask.java:
```diff
@@ -0,0 +1,231 @@
+      Object[] futures = new Object[rowMutationsList.size()];
+      table.batch(rowMutationsList, futures);
+
+      // Make sure that there were no errors returned via futures.
```
Yup. renamed.

vermas2012 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/CloudBigtableReplicator.java:
```diff
@@ -0,0 +1,305 @@
+  // batch capacity by counts and bytes. These capacity are set by `replication.source.nb.capacity`
+  // and `replication.source.size.capacity` config keys.
+  public static final long DEFAULT_BATCH_SIZE_IN_BYTES = 500_000;
+  private static long batchSizeThresholdInBytes;
```
no reason actually. fixed it.

vermas2012 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/CloudBigtableReplicator.java:
```diff
@@ -0,0 +1,305 @@
+    executorService =
+        new ThreadPoolExecutor(
+            numThreads,
+            numThreads,
```
The replication tasks are "short lived" tasks, if there is no load on the replication, the threads can be shut down. I made this decision when my threadpool was not static. Now that its shared by all instances, I don't think there is much harm in making it FixedThreadPool.

vermas2012 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/CloudBigtableReplicator.java:
```diff
@@ -0,0 +1,305 @@
+      String projectId = configuration.get(PROJECT_KEY);
+      String instanceId = configuration.get(INSTANCE_KEY);
+      // If an App profile is provided, it will be picked automatically by the connection.
+      connection = BigtableConfiguration.connect(configuration);
```
I don't think if HBase re-uses this object, but it makes sense to be safe and copy it.

vermas2012 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/README.md:
````diff
@@ -0,0 +1,31 @@
+## Properties
+Have the below properties set in `hbase-site.xml` and add it to the HBase region server classpath.
+```
+<property> 
+    <name>google.bigtable.project.id</name>
+    <value>PROJECT_KEY</value>
+    <description>
+       Cloud Bigtable Project ID
+    </description>
+</property>
+<property>
+    <name>google.bigtable.instance.id</name>
+    <value>INSTANCE_KEY</value>
+    <description>
+       Cloud Bigtable Instance ID
+    </description>
````
I think we should keep it the same. Allowing people to use same hbase-site.xml for application and replication servers.

vermas2012 commented:
Thanks for the thorough review :), I was trying to unblock the customer, but I guess we did the final review. 

PTAL.

igorbernstein2 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/CloudBigtableReplicator.java:
```diff
@@ -0,0 +1,289 @@
+            numThreads,
+            new ThreadFactoryBuilder()
+                .setDaemon(true)
+                .setNameFormat("cloud-bigtable-replication-sink-")
```
I think you want this to be `cloud-bigtable-replication-sink-%d`

igorbernstein2 commented on hbase-migration-tools/bigtable-hbase-replication/pom.xml:
```diff
@@ -0,0 +1,40 @@
+  <parent>
+    <groupId>com.google.cloud.bigtable</groupId>
+    <artifactId>hbase-migration-tools</artifactId>
+    <version>2.0.0-beta7-SNAPSHOT</version> <!-- {x-version-update:bigtable-client-parent:current} -->
```
this looks wrong

igorbernstein2 commented on hbase-migration-tools/pom.xml:
```diff
@@ -0,0 +1,38 @@
+  <parent>
+    <groupId>com.google.cloud.bigtable</groupId>
+    <artifactId>bigtable-client-parent</artifactId>
+    <version>2.0.0-beta7-SNAPSHOT</version> <!-- {x-version-update:bigtable-client-parent:current} -->
```
this looks wrong

igorbernstein2 commented on hbase-migration-tools/pom.xml:
```diff
@@ -0,0 +1,38 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <groupId>com.google.cloud.bigtable</groupId>
+  <version>2.0.0-beta7-SNAPSHOT</version> <!-- {x-version-update:bigtable-client-parent:current} -->
```
this looks wrong

vermas2012 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-1.x-replication/pom.xml:
```diff
@@ -0,0 +1,135 @@
+    <artifactId>bigtable-hbase-replication</artifactId>
+    <groupId>com.google.cloud.bigtable</groupId>
+    <version>0.0.1-SNAPSHOT
+    </version> <!-- {x-version-update:bigtable-hbase-replication:current} -->
```
Formatters keep moving into different lines, I have the googler formatter set on my intellij, it doesn't help :(

kolea2 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/utils/TestUtils.java:
```diff
@@ -0,0 +1,217 @@
+  public static final byte[] VALUE = "Value".getBytes();
+  public static final long TIMESTAMP = 1000l;
+
+  // ONLY STATIC METHODS
```
do we need this comment anymore?

vermas2012 replied:
anymore? Nothing changed, this is a utility class with static method, we explicitly prohibit object creation...

kolea2 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/test/java/com/google/cloud/bigtable/hbase/replication/CloudBigtableReplicationTaskTest.java:
```diff
@@ -0,0 +1,235 @@
+
+  @Test
+  public void batchCallFailsPartially() throws IOException, InterruptedException {
+    doThrow(new RetriesExhaustedWithDetailsException("Dummy error"))
```
please replace as "placeholder error"

kolea2 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-1.x-replication/pom.xml:
```diff
@@ -0,0 +1,135 @@
+  <parent>
+    <artifactId>bigtable-hbase-replication</artifactId>
+    <groupId>com.google.cloud.bigtable</groupId>
+    <version>0.0.1-SNAPSHOT
```
asking because - please double check when the release PR is created that it will release at the version you are expecting

vermas2012 replied:
we are not planning to release it, yet. For now, we are going to give it as "pre-release" 0.0.1 is fine for that.

vermas2012 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-2.x-replication/pom.xml:
```diff
@@ -0,0 +1,139 @@
+    <build>
+        <plugins>
+            <plugin>
+                <artifactId>maven-compiler-plugin</artifactId>
```
we discussed, the google common setting was enforcing jdk7.

kolea2 replied:
where? it should be inheriting from the shared config (https://github.com/googleapis/java-bigtable-hbase/blob/main/pom.xml#L34), which is setting it to java 8 (https://github.com/googleapis/java-shared-config/blob/main/pom.xml#L169-L181)

vermas2012 replied:
Personally, I am ok either ways. I think we are using Java8 code through out the hbase java client, so whatever works for client will work on the replication package

kolea2 replied:
that is for the enforcer plugin, not `maven-compiler-plugin`. Unless I'm missing something, I think this can and should be deleted.

vermas2012 replied:
done

vermas2012 commented on README.md:
```diff
@@ -46,7 +46,7 @@ which makes it easy for development teams to get started.
     <groupId>com.google.cloud.bigtable</groupId>
     <artifactId>bigtable-hbase-1.x</artifactId>
-    <version>2.0.0-beta6</version>
+    <version>2.0.1-SNAPSHOT</version>
```
why?  the readme should have commands to install the latest version, shouldn't it? SNAPSHOT is probably not required, but why now 2.0.1?

vermas2012 replied:
never mind, I see it is autoupdated.

kolea2 reviewed (dismissed):
for some reason I can't leave a comment on the file itself - but it looks like this was accidental? Added: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/ChannelPool.java

vermas2012 commented:
> for some reason I can't leave a comment on the file itself - but it looks like this was accidental? Added: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/ChannelPool.java

I don't know how it got committed. Removed.

vermas2012 commented:
@igorbernstein2 whenever you are ready.

igorbernstein2 commented on hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-1.x-replication/pom.xml:
```diff
@@ -0,0 +1,175 @@
+      <groupId>com.google.cloud.bigtable</groupId>
+      <artifactId>bigtable-hbase-replication-core</artifactId>
+      <version>0.0.1-SNAPSHOT
+      </version> <!-- {x-version-update:bigtable-hbase-replication:current} -->
```
please make sure to keep the comment on the same line as the version

vermas2012 replied:
done. Checked in all the new POMs.

vermas2012 replied:
I double checked all the new POMs. all good.
